### PR TITLE
Cosmic time bug

### PR DESCRIPTION
**Pull request by rstoneback**
# Description
Fixes non-unique data and file times in COSMIC

Addresses # 363, 364

There can be multiple profiles or files for a given time. The cosmic_gps code was 
updated to provide unique times.

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide
instructions so we can reproduce. Please also list any relevant details for
your test configuration

Added unique time checks within the code itself to self-monitor for issues. Tested
on multiple days and file types, locally.

**Test Configuration**:
* macOS 10.15.3

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``master``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

**Comment by jklenzing** [comment]
Running on python 3.7.3, pandas 0.24.0:
```
AttributeError: 'DataFrame' object has no attribute 'fileStamp'
```

**Comment by jklenzing** [comment]
See  changes to travis.yml in #365 for fix to python 3.7 tests in Travis CI.

**Comment by jklenzing** [comment]
scnlv1 files have a different variables, which means this fix is not universal.  A unique fix for scnlv1 could make use of `antenna_id` and `prn_id`, but these aren't in ionprf.

**Comment by rstoneback** [comment]
Thanks Jeff. I thought I checked that file type.


On Feb 8, 2020, at 10:39 AM, Jeff Klenzing <notifications@github.com<mailto:notifications@github.com>> wrote:


scnlv1 files have a different variables, which means this fix is not universal. A unique fix for scnlv1 could make use of antenna_id and prn_id, but these aren't in ionprf.

**Comment by jklenzing** [comment]
Need to compare pysat load with generalized netcdf load to ensure all variables are copied correctly

**Comment by rstoneback** [comment]
For sonprf the variable that is causing the problem is MSL_alt

float64 MSL_alt(MSL_alt)
    long_name: Mean Sea Level geometric height
    units: km
    missing_value: -999
    _FillValue: -999.0
unlimited dimensions: 
current shape = (41,)
filling off

Best guess is it doesn't like that missing value is an integer while the variable itself is a float64.

The data is not all being loaded. This appears to be a netCDF library issue or a file issue. I don't think there is anything we can do other than perhaps report this to netCDF4 or perhaps UCAR.

**Comment by aburrell** [comment]
Can we cast the missing and fill values as the specified type (float64 in this case), to ensure we are immune to this bug in file creation?

**Comment by rstoneback** [comment]
> Can we cast the missing and fill values as the specified type (float64 in this case), to ensure we are immune to this bug in file creation?

I don't currently know of a way. Everything is happening within netCDF. If I try to get the data at all using data.variables[key][:] it throws up the warning and does not provide the data. I can't change the fill or missing value attributes either, or at least I don't know how.

data.variables[key].missing_value = -999.
*** AttributeError: NetCDF: Write to read only

I thought the missing value was an int but it is actually a unicode string.
data.variables[key].missing_value
u'-999'

**Comment by jklenzing** [comment]
It sounds like this is an error that should be reported to UCAR.  Warning messages already propagate through the code to the end user.

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
   - Fixed a bug where `remote_file_list` would fail for some instruments.
   - Made import of methods more robust
   - Fixed `SettingWithCopyWarning` in `cnofs_ivm` cleaning routine
+  - Added time mangling to ensure COSMIC files and data have unique times
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply CHANGELOG.md]: I updated my description

**Comment by rstoneback** [comment]
I agree. I'm on a telecon with UCAR in about an hour and I'll bring it up to them.

**Comment by rstoneback** [comment]
I will update as I hear things. In the meantime, what do we do here? Add a note to the cosmic docstring and merge things in? Nominal update to warnings committed.

**Comment by jklenzing** [comment]
I say add a note to the docstring.  The bug fixes we have need to get merged in, and the other ones are out of our control.

**Comment by rstoneback** [comment]
Test failures are related to netCDF4 hdf errors, unrelated to changes here.

**Pull request merged**